Created yaml document which explains and exposes failures of current …b

yeatersink opened the pull request:
…HBO table.

bertfrees commented:
Perfect.

bertfrees reviewed (approved):
I added the necessary xfail flags.

egli commented on tests/braille-specs/hbo.yaml:
```diff
@@ -24,3 +46,36 @@ tests:
+  - ["אֱלֹהִים ", "⠁⠢⠇⠕⠓⠔⠍ ", xfail: true]
+  - ["הָאָרֶץ.", "⠓⠣⠁⠣⠗⠑⠋⠮", xfail: true]
+  - ["הָיְתָה תֹה. ", "⠓⠣⠚⠄⠹⠣⠓ ⠹⠕⠓⠲ ", xfail: true]
+  - ["וְהָאָרֶץ ‬", "⠺⠄⠓⠣⠁⠣⠗⠑⠮ ‬", xfail: true]
```
What is this thing about hidden Unicode characters ("This file contains bidirectional Unicode text that may be interpreted or compiled differently than what appears below. To review, open the file in an editor that reveals hidden Unicode characters", [learn more](https://github.blog/changelog/2021-10-31-warning-about-bidirectional-unicode-text/)). Apparently the problem is the U+202C character.

bertfrees replied:
Yes, I noticed that too, and when I deleted it, both in the input and output, I got an unexpected pass.

yeatersink commented:
@egli  and @bertfrees  that character is a hidden unicode character that has crept in from Logos Bible software.  They use characters like that in their formatting. They are usually stripped out in MS Word when we a do a "paste special" and remove  those annoying formatting characters.

Several of us blind folks use Logos to export our original language material as a MS Word document, then convert that Word doc in to braille. 

I forgot to make sure that the character was removed before I pushed that yaml file. That was the same problem with the yaml test in the previous pull request. Once I removed it it was better. 

If I need to remove the character, run it again, and push the updated yaml file, I will.

bertfrees commented:
@yeatersink I assume it's the same with the U+200C character (zero-width non-joiner). If we remove both of these special characters from the tests, we get two tests that already pass without your latest table changes. That's not really a problem though.

But there's another issue: after applying your latest table changes from https://github.com/liblouis/liblouis/pull/1614, some tests that used to pass fail, and some tests that used to fail are not fixed. These are the tests that used to pass but fail after https://github.com/liblouis/liblouis/pull/1614:

Input:    'אֱלֹהֵ֖ינוּ'
Expected: '⠁⠢⠇⠸⠪⠓⠌⠨⠞⠚⠝⠥'
Received: '⠁⠢⠇⠕⠓⠌⠨⠞⠚⠝⠥'

Input:    'וְאָ֣הַבְתָּ֔'
Expected: '⠺⠄⠁⠣⠨⠥⠓⠉⠧⠄⠹⠣⠐⠂'
Received: '⠺⠄⠁⠣⠨⠥⠓⠉⠧⠄⠳⠣⠂'

Input:    'אֱלֹהֶ֑יךָ'
Expected: '⠁⠢⠇⠸⠪⠓⠑⠆⠚⠡⠣'
Received: '⠁⠢⠇⠕⠓⠑⠆⠚⠡⠣'

Input:    'וּבְכָל־מְאֹדֶֽךָ׃'
Expected: '⠥⠧⠄⠡⠣⠇⠤⠍⠄⠁⠸⠪⠙⠑⠈⠡⠣⠲'
Received: '⠥⠧⠄⠡⠣⠇⠤⠍⠄⠁⠕⠙⠑⠈⠡⠣⠲'

And this is the test that https://github.com/liblouis/liblouis/pull/1614 did not fix:

Input:    'הָאָרֶץ.'
Expected: '⠓⠣⠁⠣⠗⠑⠋⠮'
Received: '⠓⠣⠁⠣⠗⠑⠮⠲'

Any idea?

yeatersink commented:
@bertfrees  Yes i do know why. 

The examples that you are providing are actually producing perfect braille. I have read your examples of what Lib Louis is producing now, and every example that you are providing is exactly and perfect of what they should be now. 

You gave 4 examples the first three are all examples of of the same issue of a holum which in the old table was represented by dots 456 and 246. In 1614, I updated that character to produce dots 145 in alignment with IHBC. This is because the He Il computer braille table does the same,  and it made sense to have consistency with this table, the He Il computer  braille table, and the IHBC. The braille is actually cleaner and makes more sense in context as well. 

the fourth example that you gave is actually a typo I believe. In this yaml file in 1625, the same input""הָאָֽרֶץ׃",produces the same braille as expected in the example you provided, by producing  "⠓⠣⠁⠣⠈⠗⠑⠮⠲"   

I am over the moon that this updated table in 1614 is producing the best and most accurate braille. I just checked 1614 to see if I need to update the YAML in 1614, but the YAML in 1614 and 1625 are both correct. They are both correct in the braille that they are expecting, and if 1614 is producing the braille that you provided above,then 1614 is producing the correct braille as well. 

I actually just checked the Common Characters file in 1614 and found that the final tsadi
with a period was represented by dots 124, 2346. This is exactly what you said was what was expected in your fourth example. This is a typo in the table.  

I have corrected the table and will push those changes to 1614 immediately. 

If you want me to, I will also update the YAML in 1625 to contain a common about the change in the Holem from 456, 246 to 135 if you want me to.

yeatersink commented:
@bertfrees I just checked the He common characters yaml and found the typo in the yaml file as well. I just updated the yaml file and the Table to reflect the typo change from 124, 2346 to the correct 2346,256.  Pushing those changes now.

bertfrees commented:
Hi @yeatersink 

> I actually just checked the Common Characters file in 1614 and found that the final tsadi
with a period was represented by dots 124, 2346. This is exactly what you said was what was expected in your fourth example. This is a typo in the table.
>
> I have corrected the table and will push those changes to 1614 immediately.

Thanks. That fixed it.

The bad news is that your changes did not fix the other failures, in fact we've got three new ones. In total we've got 7 failures now:

- Input:    'שְׁמַ֖עa'
  Expected: '⠩⠄⠍⠉⠨⠞⠫⠁'
  Received: '⠩⠄⠍⠉⠨⠞⠐⠫⠁'
  
- Input:    'יִשְׂרָאֵ֑ל'
  Expected: '⠚⠊⠱⠄⠗⠣⠁⠌⠆⠇'
  Received: '⠔⠱⠄⠗⠣⠁⠌⠆⠇'
  
- Input:    'אֱלֹהֵ֖ינוּ'
  Expected: '⠁⠢⠇⠸⠪⠓⠌⠨⠞⠚⠝⠥'
  Received: '⠁⠢⠇⠕⠓⠌⠨⠞⠚⠝⠥'
  
- Input:    'וְאָ֣הַבְתָּ֔'
  Expected: '⠺⠄⠁⠣⠨⠥⠓⠉⠧⠄⠹⠣⠐⠂'
  Received: '⠺⠄⠁⠣⠨⠥⠓⠉⠧⠄⠳⠣⠂'
  
- Input:    'אֱלֹהֶ֑יךָ'
  Expected: '⠁⠢⠇⠸⠪⠓⠑⠆⠚⠡⠣'
  Received: '⠁⠢⠇⠕⠓⠑⠆⠚⠡⠣'
  
- Input:    'וּבְכָל־מְאֹדֶֽךָ׃'
  Expected: '⠥⠧⠄⠡⠣⠇⠤⠍⠄⠁⠸⠪⠙⠑⠈⠡⠣⠲'
  Received: '⠥⠧⠄⠡⠣⠇⠤⠍⠄⠁⠕⠙⠑⠈⠡⠣⠲'
  
- Input:    'הַמָּיִם.'
  Expected: '⠓⠉⠐⠍⠣⠚⠊⠍⠲'
  Received: '⠓⠉⠐⠍⠣⠔⠍⠲'

When you said:

> The examples that you are providing are actually producing perfect braille. I have read your examples of what Lib Louis is producing now, and every example that you are providing is exactly and perfect of what they should be now.

did you mean that the _expected_ braille is correct, or the _received_ braille is correct? In the former case, the table is correct, but the test is wrong. In the latter case, the test is correct but the table contains mistakes. In any case, the expected braille doesn't match the received braille, so there is something wrong.

By the way, your last commit actually broke the table, in that it did not compile anymore. I was able to make it compile again by replacing a bunch of `letter` rules to `always`. So now the errors are gone, but we have the 7 test failures mentioned above.

yeatersink commented:
Yes, sir, the test is in accurate but the Braille that Lib Luis is producing now is correct  all of the errors look like they're from the very first test except for the last error.   Lib Luis is producing the correct Braille. It is the test that are wrong. And I think this should be expected because the errors are from the first test. There's been several changes to the table that updated some characters that are throwing those contradictions between the first test and the current    Test. Matt Yeater- Yeshurun On Sep 2, 2024, at 1:45 PM, Bert Frees ***@***.***> wrote:﻿
Hi @yeatersink

I actually just checked the Common Characters file in 1614 and found that the final tsadi
with a period was represented by dots 124, 2346. This is exactly what you said was what was expected in your fourth example. This is a typo in the table.
I have corrected the table and will push those changes to 1614 immediately.

Thanks. That fixed it.
The bad news is that your changes did not fix the other failures, in fact we've got three new ones. In total we've got 7 failures now:


Input:    'שְׁמַ֖עa'
Expected: '⠩⠄⠍⠉⠨⠞⠫⠁'
Received: '⠩⠄⠍⠉⠨⠞⠐⠫⠁'


Input:    'יִשְׂרָאֵ֑ל'
Expected: '⠚⠊⠱⠄⠗⠣⠁⠌⠆⠇'
Received: '⠔⠱⠄⠗⠣⠁⠌⠆⠇'


Input:    'אֱלֹהֵ֖ינוּ'
Expected: '⠁⠢⠇⠸⠪⠓⠌⠨⠞⠚⠝⠥'
Received: '⠁⠢⠇⠕⠓⠌⠨⠞⠚⠝⠥'


Input:    'וְאָ֣הַבְתָּ֔'
Expected: '⠺⠄⠁⠣⠨⠥⠓⠉⠧⠄⠹⠣⠐⠂'
Received: '⠺⠄⠁⠣⠨⠥⠓⠉⠧⠄⠳⠣⠂'


Input:    'אֱלֹהֶ֑יךָ'
Expected: '⠁⠢⠇⠸⠪⠓⠑⠆⠚⠡⠣'
Received: '⠁⠢⠇⠕⠓⠑⠆⠚⠡⠣'


Input:    'וּבְכָל־מְאֹדֶֽךָ׃'
Expected: '⠥⠧⠄⠡⠣⠇⠤⠍⠄⠁⠸⠪⠙⠑⠈⠡⠣⠲'
Received: '⠥⠧⠄⠡⠣⠇⠤⠍⠄⠁⠕⠙⠑⠈⠡⠣⠲'


Input:    'הַמָּיִם.'
Expected: '⠓⠉⠐⠍⠣⠚⠊⠍⠲'
Received: '⠓⠉⠐⠍⠣⠔⠍⠲'


When you said:

The examples that you are providing are actually producing perfect braille. I have read your examples of what Lib Louis is producing now, and every example that you are providing is exactly and perfect of what they should be now.

did you mean that the expected braille is correct, or the received braille is correct? In the former case, the table is correct, but the test is wrong. In the latter case, the test is correct but the table contains mistakes. In any case, the expected braille doesn't match the received braille, so there is something wrong.
By the way, your last commit actually broke the table, in that it did not compile anymore. I was able to make it compile again by replacing a bunch of letter rules to always. So now the errors are gone, but we have the 7 test failures mentioned above.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

bertfrees commented:
@yeatersink Can you please update all the tests so that they contain the correct expected braille?

Regarding whether this first batch of tests are expected to be failing: I didn't assume that, actually I assumed the opposite. The only improvement to the table (apart from the refactoring) that I am aware of is that it now correctly handles different stanza orders. Since the first tests did not demonstrate the problem with the consonant-vowel-dagesh stanza (going by the comment), I would expect they are not affected by the change.

That brings me to the other thing I wanted to ask you: we need a short summary of the improvements to the hbo table in this release that we could include in the release notes? There is the stanza order thing. What else?

yeatersink commented:
Yes, sir, I'm out right now, but I can be home in an hour or so. Please forgive me, I misunderstood. I thought that it was OK for this test to show errors to show the differences in the table. It would've literally taken me two minutes to update this test. I misunderstood what the expectation was.Updated modifications to the table:Included common characters for IHBC standard for all Hebrew. Updated Braille character for Vav. Repaired orderings of stanzas to ensure accurate braille for letters with dagesh. Corrected "final" letters for input with Braille display. Provided support for back translation.  On Sep 2, 2024, at 2:44 PM, Bert Frees ***@***.***> wrote:﻿
@yeatersink Can you please update all the tests so that they contain the correct expected braille?
Regarding whether this first batch of tests are expected to be failing: I didn't assume that, actually I assumed the opposite. The only improvement to the table (apart from the refactoring) that I am aware of is that it now correctly handles different stanza orders. Since the first tests did not demonstrate the problem with the consonant-vowel-dagesh stanza (going by the comment), I would expect they are not affected by the change.
That brings me to the other thing I wanted to ask you: we need a short summary of the improvements to the hbo table in this release that we could include in the release notes? There is the stanza order thing. What else?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

egli commented:
@yeatersink sorry I have to start making the release and there isn't enough time to work this out. Hopefully we can include it in the December release

bertfrees commented:
@yeatersink If you have time it would be good to try to finish this table now, so that it will definitely be ready for the December release. It is still fresh on my mind now, and I think we were almost there.

yeatersink commented:
Yes sir it's fresh on my mind too. I have everything pretty much complete. Just need to push the changes. Will likely do this September 13Matt Yeater- Yeshurun On Sep 12, 2024, at 10:35 AM, Bert Frees ***@***.***> wrote:﻿
@yeatersink If you have time it would be good to try to finish this table now, so that it will definitely be ready for the December release. It is still fresh on my mind now, and I think we were almost there.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

paulGeoghegan commented:
@bertfrees and @egli do you have any feedback on this as well?

bertfrees commented:
Sorry, I hadn't noticed the new commit! Thanks.

The good news is that we have more test data now.

But the bad news is that there are still many problems, in fact more problems than we had before. Don't get me wrong, the table has probably improved, but that isn't reflected in the tests:

- When I run the tests from this PR against the https://github.com/liblouis/liblouis/pull/1614 version of the table, a lot of the first batch of tests start failing. There are 7 that fail now.
- Not all tests are fixed by https://github.com/liblouis/liblouis/pull/1614. I'm not saying that all tests need to pass, but there sure are a lot of failures, which may be an indication that something is wrong, either in the table, or in the test data.

Do you actually run the tests? That would be enormously helpful. I wouldn't have to tell you about the test failures, you would notice it yourself.

By the way, these last commits added a bit of a mess. They introduced a lot of duplicate tests. Not a huge issue, but I removed them because it was obviously a mistake.

yeatersink commented:
I'm going to email you off list because I'm very confused as to why they keepsI'm going to email you off list  Matt Yeater- Yeshurun On Oct 9, 2024, at 9:10 PM, Bert Frees ***@***.***> wrote:﻿
Sorry, I hadn't noticed the new commit! Thanks.
The good news is that we have more test data now.
But the bad news is that there are still many problems, in fact more problems than we had before. Don't get me wrong, the table has probably improved, but that isn't reflected in the tests:

When I run the tests from this PR against the #1614 version of the table, a lot of the first batch of tests start failing. There are 7 that fail now.
Not all tests are fixed by #1614. I'm not saying that all tests need to pass, but there sure are a lot of failures, which may be an indication that something is wrong, either in the table, or in the test data.

Do you actually run the tests? That would be enormously helpful. I wouldn't have to tell you about the test failures, you would notice it yourself.
By the way, these last commits added a bit of a mess. They introduced a lot of duplicate tests. Not a huge issue, but I removed them because it was obviously a mistake.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

paulGeoghegan commented:
@bertfrees just to clarify you are running this test against the hbo.utb file from that new PR right?
We have been trying to figure this out and still haven't been able to get LIblouis working anywhere.

bertfrees commented:
Incorporated in https://github.com/liblouis/liblouis/pull/1676
